src/Simulation/Common/IQuantumProcessor.cs

```diff
@@ -0,0 +1,587 @@
+        /// <remarks>
+        /// </remarks>
+        void OnFail(System.Runtime.ExceptionServices.ExceptionDispatchInfo exceptionDispatchInfo);
+
+        /// <summary>
+        /// Called when qubits are allocated by Q# <a href="https://docs.microsoft.com/quantum/language/statements#clean-qubits"><c>using</c></a> block. 
+        /// </summary>
+        /// <param name="qubits">Qubits that are being allocated</param>.
+        /// <remarks>
+        /// Every qubit has a unique identifier <see cref="Qubit.Id"/>.
+        /// All newly allocated qubits are in |0⟩ state.
+        /// </remarks>
+        void OnAllocateQubits(IQArray<Qubit> qubits);
+
+        /// <summary>
+        /// Called when qubits are released in Q# in the end of <a href="https://docs.microsoft.com/quantum/language/statements#clean-qubits"><c>using</c></a> block. 
+        /// </summary>
+        /// <param name="qubits">Qubits that are being released</param>.
+        /// <remarks>
+        /// Every qubit has a unique identifier <see cref="Qubit.Id"/>.
+        /// All qubits are expected to be released in |0⟩ state.
+        /// </remarks>
+        void OnReleaseQubits(IQArray<Qubit> qubits);
+
+        /// <summary>
+        /// Called when qubits are borrowed by Q# <a href="https://docs.microsoft.com/quantum/language/statements#dirty-qubits"><c>borrowing</c></a> block. 
+        /// </summary>
+        /// <param name="qubits">Qubits that are being borrowed</param>.
+        /// <remarks>
+        /// Every qubit has a unique identifier <see cref="Qubit.Id"/>.
+        /// Borrowed qubits can be in any state.
+        /// </remarks>
+        void OnBorrowQubits(IQArray<Qubit> qubits);
+
+        /// <summary>
+        /// Called when qubits are returned in the end of Q# <a href="https://docs.microsoft.com/quantum/language/statements#dirty-qubits"><c>borrowing</c></a> block. 
+        /// </summary>
+        /// <param name="qubits">Qubits that have been borrowed and are now being returned</param>.
+        /// <remarks>
+        /// Every qubit has a unique identifier <see cref="Qubit.Id"/>.
+        /// Borrowed qubits are expected to be returned in the same state as the state they have been borrowed in.
+        /// </remarks>
+        void OnReturnQubits(IQArray<Qubit> qubits);
+
+        /// <summary>
+        /// Called when 
+        /// <a href="https://docs.microsoft.com/qsharp/api/qsharp/microsoft.quantum.diagnostics.dumpmachine">Microsoft.Quantum.Diagnostics.DumpMachine</a>
+        /// is called in Q#.
+        /// </summary>
+        /// <remarks>
+        /// The names and the order of the parameters are similar to corresponding Q# operation. 
+        /// </remarks>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="location">Provides information on where to generate the DumpMachine state.</param>
+        void OnDumpMachine<T>(T location);
```

Comment:
why is it necessary to provide/implement  `OnDump` handlers?

Reply:
This is part of our standard simulator interface. Useful for debugging and for research on what the machine's state is. Not every physical machine will be able to dump, but some may, especially simulators.

---
In reply to: [351129532](https://github.com/microsoft/qsharp-runtime/pull/82#discussion_r351129532) [](ancestors = 351129532)